File: examples/python/agg-hand/handlers/reveal_cards.py
"""Handler for RevealCards command."""

import sys
from pathlib import Path
from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

sys.path.insert(0, str(Path(__file__).parent.parent.parent / "angzarr"))
sys.path.insert(0, str(Path(__file__).parent.parent))

from angzarr_client.proto.angzarr import types_pb2 as types
from angzarr_client.errors import CommandRejectedError
from angzarr_client.proto.examples import hand_pb2 as hand
from angzarr_client.proto.examples import types_pb2 as poker_types

from .state import HandState
from .game_rules import get_game_rules


def handle_reveal_cards(
    command_book: types.CommandBook,
    command_any: Any,
    state: HandState,
    seq: int,
) -> types.EventBook:
    """Handle RevealCards command."""
    if not state.exists():
        raise CommandRejectedError("Hand not dealt")

    if state.status != "showdown":
        raise CommandRejectedError("Not in showdown")

    cmd = hand.RevealCards()
    command_any.Unpack(cmd)

    if not cmd.player_root:
        raise CommandRejectedError("player_root is required")

    # Find the player
    player = None
    for p in state.players.values():
        if p.player_root == cmd.player_root:
            player = p
            break

    if not player:
        raise CommandRejectedError("Player not in hand")

    if player.has_folded:
        raise CommandRejectedError("Player has folded")

    if cmd.muck:
        # Player chooses to muck
        event = hand.CardsMucked(
            player_root=cmd.player_root,
            mucked_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
        )
    else:
        # Reveal cards and evaluate hand
        rules = get_game_rules(state.game_variant)
        rank_type, score, kickers = rules.evaluate_hand(
            player.hole_cards,
            state.community_cards,
        )

        event = hand.CardsRevealed(
            player_root=cmd.player_root,
            ranking=poker_types.HandRanking(
                rank_type=rank_type,
                kickers=[k for k in kickers],
                score=score,
            ),
            revealed_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
        )

        for suit, rank in player.hole_cards:
            event.cards.append(poker_types.Card(suit=suit, rank=rank))

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.poker/")

    return types.EventBook(
        cover=command_book.cover,
        pages=[
            types.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(
                    seconds=int(datetime.now(timezone.utc).timestamp())
                ),
            )
        ],
    )

File: examples/rust/agg-hand/src/handlers/reveal_cards.rs
//! RevealCards command handler.

use angzarr_client::proto::examples::{CardsMucked, CardsRevealed, HandRanking, RevealCards};
use angzarr_client::proto::{event_page, CommandBook, EventBook, EventPage};
use angzarr_client::{pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::game_rules::get_rules;
use crate::state::HandState;

pub fn handle_reveal_cards(
    command_book: &CommandBook,
    command_any: &Any,
    state: &HandState,
    seq: u32,
) -> CommandResult<EventBook> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Hand does not exist"));
    }
    if state.is_complete() {
        return Err(CommandRejectedError::new("Hand already complete"));
    }

    let cmd: RevealCards = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    let player = state
        .get_player(&cmd.player_root)
        .ok_or_else(|| CommandRejectedError::new("Player not in hand"))?;

    if player.has_folded {
        return Err(CommandRejectedError::new("Player has folded"));
    }

    let event_any = if cmd.muck {
        let event = CardsMucked {
            player_root: cmd.player_root,
            mucked_at: Some(angzarr_client::now()),
        };
        pack_event(&event, "examples.CardsMucked")
    } else {
        // Use game rules to evaluate hand properly
        let rules = get_rules(state.game_variant);
        let hand_rank = rules.evaluate_hand(&player.hole_cards, &state.community_cards);

        let ranking = HandRanking {
            rank_type: hand_rank.rank_type as i32,
            kickers: hand_rank.kickers.into_iter().map(|r| r as i32).collect(),
            score: hand_rank.score,
        };

        let event = CardsRevealed {
            player_root: cmd.player_root,
            cards: player.hole_cards.clone(),
            ranking: Some(ranking),
            revealed_at: Some(angzarr_client::now()),
        };
        pack_event(&event, "examples.CardsRevealed")
    };

    Ok(EventBook {
        cover: command_book.cover.clone(),
        pages: vec![EventPage {
            sequence: Some(event_page::Sequence::Num(seq)),
            event: Some(event_any),
            created_at: Some(angzarr_client::now()),
        }],
        snapshot: None,
        next_sequence: 0,
    })
}
